compiler error in correct program with specific declarative macros

### Summary

In my project, I use declarative macros to better work with string IDs for GTK actions. For this, I use declarative macros declared by a second declarative macro. I found that clippy fails while trying to use `cargo clippy --fix` on it, while compiling and running it with `cargo run` works without fail.

I understand this is rather specific and an edge case, but since the output specifically says to report this, I'm opening this bug report so that it is at least reported.

<details>

<summary>Clippy Output</summary>

```
plex@theseus /tmp/tmp.kly9o3CJTt/aa % cargo run                       
   Compiling aa v0.1.0 (/tmp/tmp.kly9o3CJTt/aa)
warning: unused macro definition: `simple_action`
 --> src/main.rs:4:22
  |
4 |         macro_rules! simple_action {
  |                      ^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_macros)]` on by default

warning: unused macro definition: `A_ID_SETTINGS_DELETE_IDENTITY`
  --> src/main.rs:40:14
   |
40 |         aid!(A_ID_SETTINGS_DELETE_IDENTITY, "settings.delete_identity");
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused import: `$id`
  --> src/main.rs:31:32
   |
31 |                 pub(crate) use $id;
   |                                ^^^
...
40 |         aid!(A_ID_SETTINGS_DELETE_IDENTITY, "settings.delete_identity");
   |         --------------------------------------------------------------- in this macro invocation
   |
   = note: `#[warn(unused_imports)]` on by default
   = note: this warning originates in the macro `aid` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: `aa` (bin "aa") generated 3 warnings (run `cargo fix --bin "aa"` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.11s
     Running `target/debug/aa`
Hello, world! settings.delete_everything
plex@theseus /tmp/tmp.kly9o3CJTt/aa % cargo clippy --fix --allow-dirty
    Checking aa v0.1.0 (/tmp/tmp.kly9o3CJTt/aa)
warning: failed to automatically apply fixes suggested by rustc to crate `aa`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0433]: failed to resolve: could not find `A_ID_SETTINGS_DELETE_EVERYTHING` in `ids`
  --> src/main.rs:47:27
   |
47 |         crate::modu::ids::A_ID_SETTINGS_DELETE_EVERYTHING!()
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ could not find `A_ID_SETTINGS_DELETE_EVERYTHING` in `ids`

warning: unused macro definition: `simple_action`
 --> src/main.rs:4:22
  |
4 |         macro_rules! simple_action {
  |                      ^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_macros)]` on by default

warning: unused macro definition: `A_ID_SETTINGS_DELETE_EVERYTHING`
  --> src/main.rs:36:13
   |
36 |             A_ID_SETTINGS_DELETE_EVERYTHING,
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused macro definition: `A_ID_SETTINGS_DELETE_IDENTITY`
  --> src/main.rs:40:14
   |
40 |         aid!(A_ID_SETTINGS_DELETE_IDENTITY, "settings.delete_identity");
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 1 previous error; 3 warnings emitted

For more information about this error, try `rustc --explain E0433`.
Original diagnostics will follow.

warning: unused macro definition: `simple_action`
 --> src/main.rs:4:22
  |
4 |         macro_rules! simple_action {
  |                      ^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_macros)]` on by default

warning: unused macro definition: `A_ID_SETTINGS_DELETE_IDENTITY`
  --> src/main.rs:40:14
   |
40 |         aid!(A_ID_SETTINGS_DELETE_IDENTITY, "settings.delete_identity");
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused import: `$id`
  --> src/main.rs:31:32
   |
31 |                 pub(crate) use $id;
   |                                ^^^
...
40 |         aid!(A_ID_SETTINGS_DELETE_IDENTITY, "settings.delete_identity");
   |         --------------------------------------------------------------- in this macro invocation
   |
   = note: `#[warn(unused_imports)]` on by default
   = note: this warning originates in the macro `aid` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: `aa` (bin "aa") generated 3 warnings (run `cargo clippy --fix --bin "aa"` to apply 1 suggestion)
warning: failed to automatically apply fixes suggested by rustc to crate `aa`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0433]: failed to resolve: could not find `A_ID_SETTINGS_DELETE_EVERYTHING` in `ids`
  --> src/main.rs:47:27
   |
47 |         crate::modu::ids::A_ID_SETTINGS_DELETE_EVERYTHING!()
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ could not find `A_ID_SETTINGS_DELETE_EVERYTHING` in `ids`

warning: unused macro definition: `simple_action`
 --> src/main.rs:4:22
  |
4 |         macro_rules! simple_action {
  |                      ^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_macros)]` on by default

warning: unused macro definition: `A_ID_SETTINGS_DELETE_EVERYTHING`
  --> src/main.rs:36:13
   |
36 |             A_ID_SETTINGS_DELETE_EVERYTHING,
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused macro definition: `A_ID_SETTINGS_DELETE_IDENTITY`
  --> src/main.rs:40:14
   |
40 |         aid!(A_ID_SETTINGS_DELETE_IDENTITY, "settings.delete_identity");
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 1 previous error; 3 warnings emitted

For more information about this error, try `rustc --explain E0433`.
Original diagnostics will follow.

warning: `aa` (bin "aa" test) generated 3 warnings (3 duplicates)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.15s
```

</details>

Playing around a bit, I was able to reproduce the error with the smaller code attached below.

Interestingly, the code fixes happen (at least partially) even if errors occur.

[This](https://github.com/PlexSheep/SREMP/actions/runs/18604053024/job/53049257435) CI log shows how this occured in my project.

### Reproducer

The following (a bit complicated) program can be compiled and runs with `cargo run` but `cargo clippy fix` fails on it.

```rs
mod modu {
    pub(crate) mod macros {

        macro_rules! simple_action {
            ($app:expr, $id:expr, $body:block) => {{
                let action = gtk::gio::SimpleAction::new($id, None);
                action.connect_activate(move |_, _| $body);
                $app.add_action(&action);
            }};
            ($app:expr, $state:expr, $app_c:ident, $state_c:ident, $id:expr, $body:block) => {{
                let $app_c = $app.clone();
                let $state_c = $state.clone();
                let action = gtk::gio::SimpleAction::new($id, None);
                action.connect_activate(move |_, _| $body);
                $app.add_action(&action);
            }};
        }
    }

    pub(crate) mod ids {
        macro_rules! aid {
            ($id:ident, $text:literal) => {
                macro_rules! $id {
                    () => {
                        $text
                    };
                    (app) => {
                        concat!("app.", $text)
                    };
                }
                pub(crate) use $id;
            };
        }

        aid!(
            A_ID_SETTINGS_DELETE_EVERYTHING,
            "settings.delete_everything"
        );
        // NOTE: Clippy only fails if there is another A_ID declared that is NOT USED
        aid!(A_ID_SETTINGS_DELETE_IDENTITY, "settings.delete_identity");
    }
}

fn main() {
    println!(
        "Hello, world! {}",
        crate::modu::ids::A_ID_SETTINGS_DELETE_EVERYTHING!()
    );
}
```

### Version

```text
rustc 1.90.0 (1159e78c4 2025-09-14)
binary: rustc
commit-hash: 1159e78c4747b02ef996e55082b704c09b970588
commit-date: 2025-09-14
host: x86_64-unknown-linux-gnu
release: 1.90.0
LLVM version: 20.1.8
```

### Additional Labels

_No response_

--- a/compiler/rustc_resolve/src/check_unused.rs
+++ b/compiler/rustc_resolve/src/check_unused.rs
@@ -486,6 +486,9 @@ impl Resolver<'_, '_> {
 
             let remove_whole_use = remove_spans.len() == 1 && remove_spans[0] == unused.item_span;
             let num_to_remove = ms.primary_spans().len();
+            // Only offer rustfix suggestions for spans that point at directly editable code.
+            let can_suggest_removal =
+                remove_spans.iter().all(|span| span.can_be_used_for_suggestions());
 
             // If we are in the `--test` mode, suppress a help that adds the `#[cfg(test)]`
             // attribute; however, if not, suggest adding the attribute. There is no way to
@@ -516,11 +519,13 @@ impl Resolver<'_, '_> {
                 unused.use_tree_id,
                 ms,
                 move |dcx, level, sess| {
-                    let sugg = if remove_whole_use {
-                        errors::UnusedImportsSugg::RemoveWholeUse { span: remove_spans[0] }
-                    } else {
-                        errors::UnusedImportsSugg::RemoveImports { remove_spans, num_to_remove }
-                    };
+                    let sugg = can_suggest_removal.then(|| {
+                        if remove_whole_use {
+                            errors::UnusedImportsSugg::RemoveWholeUse { span: remove_spans[0] }
+                        } else {
+                            errors::UnusedImportsSugg::RemoveImports { remove_spans, num_to_remove }
+                        }
+                    });
                     let test_module_span = test_module_span.map(|span| {
                         sess.downcast_ref::<rustc_session::Session>()
                             .expect("expected a `Session`")

--- a/compiler/rustc_resolve/src/errors.rs
+++ b/compiler/rustc_resolve/src/errors.rs
@@ -1742,7 +1742,7 @@ pub(crate) struct ElidedLifetimesInPaths {
 )]
 pub(crate) struct UnusedImports {
     #[subdiagnostic]
-    pub sugg: UnusedImportsSugg,
+    pub sugg: Option<UnusedImportsSugg>,
     #[help("if this is a test module, consider adding a `#[cfg(test)]` to the containing module")]
     pub test_module_span: Option<Span>,
 

--- a/tests/ui/imports/unused-import-in-macro-expansion-rustfix.fixed
+++ b/tests/ui/imports/unused-import-in-macro-expansion-rustfix.fixed
@@ -1,0 +1,22 @@
+//@ edition:2024
+//@ run-rustfix
+//@ rustfix-only-machine-applicable
+//@ check-pass
+
+mod m {
+    macro_rules! define_new_macro {
+        ($name:ident) => {
+            macro_rules! $name {
+                () => {};
+            }
+            pub(crate) use $name;
+        };
+    }
+
+    define_new_macro!(item_used);
+    define_new_macro!(item_unused);
+}
+
+fn main() {
+    m::item_used!();
+}

--- a/tests/ui/imports/unused-import-in-macro-expansion-rustfix.fixed
+++ b/tests/ui/imports/unused-import-in-macro-expansion-rustfix.fixed
@@ -1,3 +1,4 @@
+// Regression test for <https://github.com/rust-lang/rust/issues/147855>
 //@ edition:2024
 //@ run-rustfix
 //@ rustfix-only-machine-applicable

--- a/tests/ui/imports/unused-import-in-macro-expansion-rustfix.fixed
+++ b/tests/ui/imports/unused-import-in-macro-expansion-rustfix.fixed
@@ -1,5 +1,5 @@
 // Regression test for <https://github.com/rust-lang/rust/issues/147855>
-//@ edition:2024
+//@ edition:2021
 //@ run-rustfix
 //@ rustfix-only-machine-applicable
 //@ check-pass

--- a/tests/ui/imports/unused-import-in-macro-expansion-rustfix.rs
+++ b/tests/ui/imports/unused-import-in-macro-expansion-rustfix.rs
@@ -1,0 +1,22 @@
+//@ edition:2024
+//@ run-rustfix
+//@ rustfix-only-machine-applicable
+//@ check-pass
+
+mod m {
+    macro_rules! define_new_macro {
+        ($name:ident) => {
+            macro_rules! $name {
+                () => {};
+            }
+            pub(crate) use $name;
+        };
+    }
+
+    define_new_macro!(item_used);
+    define_new_macro!(item_unused);
+}
+
+fn main() {
+    m::item_used!();
+}

--- a/tests/ui/imports/unused-import-in-macro-expansion-rustfix.rs
+++ b/tests/ui/imports/unused-import-in-macro-expansion-rustfix.rs
@@ -1,3 +1,4 @@
+// Regression test for <https://github.com/rust-lang/rust/issues/147855>
 //@ edition:2024
 //@ run-rustfix
 //@ rustfix-only-machine-applicable

--- a/tests/ui/imports/unused-import-in-macro-expansion-rustfix.rs
+++ b/tests/ui/imports/unused-import-in-macro-expansion-rustfix.rs
@@ -1,5 +1,5 @@
 // Regression test for <https://github.com/rust-lang/rust/issues/147855>
-//@ edition:2024
+//@ edition:2021
 //@ run-rustfix
 //@ rustfix-only-machine-applicable
 //@ check-pass

